Generics.md

```diff
@@ -1,164 +1,386 @@
-
-## Summary
-
-* angle bracket syntax for generic variable declaration, and generic value specification.
-* declare generic variables on function declarations, e.g.: `fn foo<E>(arg: E) -> E { let e:E = arg; return e; }`
-* within an fn with a generic declaration,
-  generic variables names can be used in place of a wgsl type in both the fn declaration and fn body,
-  or in a function call expression inside the fn body.
-  The generic variable text will be replaced by the generic value text during linking.
-  So if `E` is `f32`  the linked wgsl for foo would be: `fn foo(arg: f32) { let e:f32 = arg; return e; }`.
-* Note that a linker will generate multiple copies of fn foo() in wgsl,
-  one for each unique set of generic arguments. So each fn will have a unique name.
-* generic variable values are supplied on import statements or call statements.
-
-      * import with a generic:
-        ```
-        import util/foo<f32> as foo32;
-        main() { foo32(1.0); }
-        ```
-      * or, call with a generic:
-        ```
-        foo<f32>(1.0);
-        ```
-
-* generic values supplied with imports are single world tokens (typically wgsl type names or function names), 
-  or generic variables declared on that function.
+# Generic Declarations and Generic Modules
 
-## Examples
+I would like to limit as much as possible the amount of additional syntax and semantics in our extensions.
+After all, we are writing an extension, not a different language!
+It should look and feel just like the original stuff and extensions should be straight forward, self-explanatory even.
+
+## The proposal
+
+It builds on 3 observations
+
+* WGSL *already* has support for parametric types and functions, they are called [type-generators](https://www.w3.org/TR/WGSL/#type-generator) and [function overloads](https://www.w3.org/TR/WGSL/#builtin-functions). It even has some syntax for them.
+  The issue is, they are not allowed to be declared in user code. This proposal changes this.
+* A *module* is a collection of functions, structs and associated types that share a semantic. Turns out, a WGSL file is just that. This proposal introduces *implicit modules* as files.
+* global value and alias declarations are immutable and must be initialized. What if we could override them? if so, this is a simple mechanism for module specialization.
+  What if we also could override struct and functions declarations? In fact, all declarations?
+
+It proposes two extensions: [**Generic Declarations**](#generic-declarations) and [**Generic Modules**](#generic-modules) (+ optional [**Function-as-Type**](#function-as-type) [**Module-as-Type**](#module-as-type)).
+They are orthogonal but they share the concepts of [*Type Hierarchy*](formalism-type-hierarchy) and [*Generic Type*](#formalism-generic-type) formalized below.
+We could support only one of the two and achieve the same expressivity AFAIK. But not the same ergonomics.
+
+## Formalism: Type Hierarchy
```

Comment:
For most people (not all!) I think it'd read better with an example or two before we dig into the formalism.